### PR TITLE
remote_settings: Add `Client::get_records_raw_with_options`.

### PR DESCRIPTION
**Pull request by linabutler**
For the Suggest collection, we'd like to specify some extra options:

* Fetch all records newer than a given timestamp. This is currently possible using `get_records_since`.
* Sort records oldest-to-newest. Remote Settings orders records newest-to-oldest by default, but we want to invert the order to support incremental downloading.
* Specify an optional limit for the number of records to fetch.

This commit:

* Adds a `GetItemsOptions` builder that sets filtering, sorting, and pagination options.
* Implements `Client::get_records_{raw, since}` in terms of the new `get_records_raw_with_options`.

Since Suggest will use `remote_settings::Client` directly from Rust, `GetItemsOptions` and `get_records_raw_with_options` are not UniFFIed.

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/remote_settings/src/client.rs [hunk 1]
```diff
@@ -61,22 +64,16 @@ impl Client {
...
     /// collection defined by the [ClientConfig] used to generate this [Client].
     pub fn get_records_since(&self, timestamp: u64) -> Result<RemoteSettingsResponse> {
-        let path = format!(
-            "v1/buckets/{}/collections/{}/records?_since={}",
-            &self.bucket_name, &self.collection_name, timestamp
-        );
-        let resp = self.make_request(path)?;
+        let resp = self.get_records_raw_with_options(
+            GetItemsOptions::new().gt("last_modified", timestamp.to_string()),
+        )?;
```
Comment (linabutler) [inline components/remote_settings/src/client.rs:76]: According to https://docs.kinto-storage.org/en/latest/api/1.x/filtering.html#polling-for-changes, `_since` is an alias for `gt_last_modified`. 😊

**Review by mhammond — APPROVED** [review]
Nice!

**Review by MatthewTighe — APPROVED** [review]
Looks great from my perspective, thank you for continuing to iterate on this!

File: components/remote_settings/src/client.rs [hunk 2]
```diff
@@ -271,10 +281,217 @@ struct AttachmentsCapability {
...
+                }
+            })
+            .chain({
+                // For selecting fields (https://docs.kinto-storage.org/en/latest/api/1.x/selecting_fields.html),
+                // the query pair syntax is `_fields=field1,field2`.
+                if self.fields.is_empty() {
+                    None
+                } else {
+                    Some(("_fields".into(), self.fields.join(",").into()))
+                }
```
Comment (pjenvey) [inline components/remote_settings/src/client.rs:426]: The newish `bool` method can be nice in places like this (and above)
```suggestion
                !(self.fields.is_empty())
                    .then_some(("_fields".into(), self.fields.join(",").into()))
```

Comment (linabutler) [reply components/remote_settings/src/client.rs:426]: Thanks @pjenvey! 🏆 I actually tried writing it with `.then(...)` first, like:

```rs
.chain(
    (!self.fields.is_empty()).then(|| ("_fields".into(), self.fields.join(",").into())),
)
```

...But that felt a little cramped, and I thought the extra parentheses around `(!self.fields.is_empty())` made it easy to miss the `!`, so I wrote it out explicitly. Although, with the comment and block there, maybe it's not too bad?

```rs
.chain({
    // For selecting fields (https://docs.kinto-storage.org/en/latest/api/1.x/selecting_fields.html),
    // the query pair syntax is `_fields=field1,field2`.
    (!self.fields.is_empty()).then(|| ("_fields".into(), self.fields.join(",").into()))
})
```

WDYT?

File: components/remote_settings/src/client.rs [hunk 3]
```diff
@@ -88,6 +85,20 @@ impl Client {
...
+        let path = format!(
+            "v1/buckets/{}/collections/{}/records",
+            &self.bucket_name, &self.collection_name
+        );
+        let mut url = self.base_url.join(&path)?;
+        for (name, value) in options.iter_query_pairs() {
+            url.query_pairs_mut().append_pair(&name, &value);
+        }
+        self.make_request(url)
+    }
```
[the remark this answers is not in the document]
Comment (linabutler) [reply components/remote_settings/src/client.rs:100]: Definitely, thanks for noting that; I'll get a follow-up bug on file! We could very likely expose `GetItemsOptions` directly in UDL, and put its mutable state behind a mutex, without sacrificing ergonomics or performance: we aren't expecting this to be hot code, and `parking_lot`'s mutexes have a fast path for the uncontended case also. That would be a great first bug for learning more about Rust and UniFFI!

File: components/remote_settings/src/client.rs (hunk 2 above)
Comment (pjenvey) [reply components/remote_settings/src/client.rs:426]: It's just a suggestion so feel free to ignore. I'll just note that I've used this a few times chaining onto the same `!is_empty()` method call and it already feels like a pretty common, easily identifiable pattern to me: if there's something in this collection, use it.

Comment (linabutler) [reply components/remote_settings/src/client.rs]: Perfect, thanks—it helps to know that you've used this a few times, and that it feels like a common pattern! I folded in your suggestion 😁

**Review by pjenvey — APPROVED** [review]

File: components/remote_settings/src/client.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (linabutler) [reply components/remote_settings/src/client.rs:100]: (Filed #5608).

**Review by NguyenCuong17012021 — COMMENTED** [review]
Seting

**Pull request merged**